ssh-copy-id: init at 8.4p1

midchildan opened the pull request:
Split the ssh-copy-id tool from the openssh package into a separate output. This mainly benefits macOS users because the version of openssh that is preintalled in macOS doesn't provide ssh-copy-id.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented:
Please target staging with the mass rebuild.

Also I am not sure if this is worth the benefit. The openssh package is ~6.23 MB and over 1000 other packages want to download it anyway.

midchildan commented:
The main problem with bundling ssh and ssh-copy-id into a single package is that it would overshadow the preinstalled ssh. This is undesirable on macOS because the preinstalled version is tweaked to integrate with the system keychain. Replacing that with openssh upstream would break existing workflows for many users.

midchildan commented:
I've just successfully ran nixos/tests/openssh.nix.

midchildan commented:
**Package closure size**

From 20.09:

```console
❯ nix path-info -S "$(which ssh)"
/nix/store/jln5qqzyy4laawfp5y4f4lqgiwja9n1l-openssh-8.3p1	 171975736
```

This PR:

```console
❯ nix path-info -S "$(nix-build --no-out-link -A openssh.out)"
/nix/store/lw81h47vxsinc1ayxvpbc1a2vqcgv59j-openssh-8.4p1	 111366840
❯ nix path-info -S "$(nix-build --no-out-link -A openssh.man)"
/nix/store/afvy3yac5jr616jvapa8f1h32zp9nnjj-openssh-8.4p1-man	     90808
❯ nix path-info -S "$(nix-build --no-out-link -A openssh.ssh_copy_id)"
/nix/store/w1wxr8kjfgm40m7dxnjwh1xidf4ybvvr-openssh-8.4p1-ssh_copy_id	  34288168
```

SuperSandro2000 commented:
> Replacing that with openssh upstream would break existing workflows for many users.

But why do you need copy-id then? Can't you just use the system one or add an alias for copy-id to the nix copy id one?

midchildan commented:
macOS doesn't come preinstalled with ssh-copy-id so people install through package managers.  For what it's worth, it ranks 271 on the list of popular packages in Homebrew so it's a somewhat popular package on macOS [1].

On the other hand, for the alias approach to work, the nix profile path would have to come after `/usr/bin` which I believe is uncommon as it's not the default. I think it's best if people could install ssh-copy-id directly.

[1]: https://formulae.brew.sh/analytics/install-on-request/365d/

matthewbauer commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -100,10 +103,12 @@ stdenv.mkDerivation rec {
+    mkdir -p $ssh_copy_id/bin $man/share/man/man1
+    cp contrib/ssh-copy-id $ssh_copy_id/bin/
+    chmod +x $ssh_copy_id/bin/ssh-copy-id
+    cp contrib/ssh-copy-id.1 $man/share/man/man1/
```
To avoid breaking things, can you install a symlink in $out as well? This is so that things like `lib.getBin openssh` have all of the outputs. This still allows for users to install ssh_copy_id output.

Something like:

```
$ ln -s $ssh_copy_id/bin/ssh-copy-id $out/bin/ssh-copy-id
```

midchildan replied:
Fixed.

matthewbauer commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6251,6 +6251,8 @@ in
       pam = if stdenv.isLinux then pam else null;
     };
 
+  ssh-copy-id = openssh.ssh_copy_id;
```
Note, that how nix-env works will ignore which output selected since. I think `nix profile install` may not have this same issue.

midchildan replied:
Oh my, I wasn't aware of this. Running `nix-env -f . -A ssh-copy-id` does indeed link the whole ssh package into my user profile.

I think I'll change this PR to split out the ssh-copy-id package instead of using multiple outputs, since nix-env likely be here to stay for years to come.

midchildan commented:
I've split `ssh-copy-id` from `openssh` into a separate package so that `nix-env -iA ssh-copy-id` won't also link `openssh` into the user profile.

SuperSandro2000 commented:
@midchildan Can you please resolve the merge conflict?

midchildan commented:
Rebased.

midchildan commented:
Re-rebased.

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
 , linkOpenssl ? true
 }:
 
+with lib;
+
```
```suggestion
```

Please do not user with lib over the entire file but only where it is needed.

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
+    fetchFromGitHub {
+      owner = "rapier1";
+      repo = "openssh-portable";
+      rev =
+        "hpn-KitchenSink-${replaceStrings [ "." "p" ] [ "_" "_P" ] version}";
```
```suggestion
      rev = "hpn-KitchenSink-${replaceStrings [ "." "p" ] [ "_" "_P" ] version}";
```

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
   gssapiPatch = fetchpatch {
     name = "openssh-gssapi.patch";
-    url = "https://salsa.debian.org/ssh-team/openssh/raw/debian/1%25${version}-2/debian/patches/gssapi.patch";
+    url =
+      "https://salsa.debian.org/ssh-team/openssh/raw/debian/1%25${version}-2/debian/patches/gssapi.patch";
```
```suggestion
    url = "https://salsa.debian.org/ssh-team/openssh/raw/debian/1%25${version}-2/debian/patches/gssapi.patch";
```

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
       ./ssh-keysign.patch
+    ] ++ optional withGssapiPatches (assert withKerberos; gssapiPatch);
+
+    postPatch =
+      # On Hydra this makes installation fail (sometimes?),
+      # and nix store doesn't allow such fancy permission bits anyway.
+      ''
+        substituteInPlace Makefile.in --replace '$(INSTALL) -m 4711' '$(INSTALL) -m 0711'
+      '';
```
```suggestion
    # On Hydra this makes installation fail (sometimes?),
    # and nix store doesn't allow such fancy permission bits anyway.
    postPatch = ''
      substituteInPlace Makefile.in --replace '$(INSTALL) -m 4711' '$(INSTALL) -m 0711'
    '';
```

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
+
+    nativeBuildInputs = [ pkg-config ]
+      ++ optional (hpnSupport || withGssapiPatches) autoreconfHook
+      ++ optional withKerberos pkgs.kerberos.dev;
```
```suggestion
      ++ optional withKerberos kerberos.dev;
```

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
+      (if pam != null then "--with-pam" else "--without-pam")
+    ] ++ optional (etcDir != null) "--sysconfdir=${etcDir}"
+      ++ optional withFIDO "--with-security-key-builtin=yes"
+      ++ optional withKerberos (assert kerberos != null; "--with-kerberos5=${kerberos}")
```
```suggestion
      ++ optional withKerberos "--with-kerberos5=${kerberos}"
```

SuperSandro2000 commented on pkgs/tools/networking/openssh/default.nix:
```diff
@@ -18,124 +19,148 @@
+  ssh-copy-id = stdenv.mkDerivation {
+    pname = "ssh-copy-id";
+    inherit src version;
+
+    phases = [ "unpackPhase" "patchPhase" "installPhase" "fixupPhase" ];
```
```suggestion
```

Please disable phases which should not run with ``dont... = true;``

midchildan replied:
Fixed.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6729,12 +6729,12 @@ in
+  inherit (callPackages ../tools/networking/openssh {
+    hpnSupport = false;
+    etcDir = "/etc/ssh";
+    pam = if stdenv.isLinux then pam else null;
```
This is not the cleanest way because packages shouldn't be overwritten with null. I would prefer a withPam or enablePam flag.

midchildan replied:
I added an `withPam` option.

dasJ commented:
Sorry for causing new conflicts to this, but I updated and refactored OpenSSH (due to that one CVE).
Why don't you just
```nix
runCommand "ssh-copy-id-${openssh.version}" ''
  mkdir -p "$out/bin"
  ln -s "${openssh}/bin/ssh-copy-id" "$out/bin"
''
```
instead of writing the whole package?

midchildan committed:
ssh-copy-id: init at 8.4p1

midchildan commented:
I've changed this PR to use `runCommand` instead following the suggestion from @dasJ. I also changed the target branch to `master` because no rebuilds would happen with this approach.

SuperSandro2000 commented on pkgs/tools/networking/openssh/copyid.nix:
```diff
@@ -0,0 +1,11 @@
+  };
+} ''
+  install -Dm 755 {${openssh},$out}/bin/ssh-copy-id
+  install -Dm 644 {${openssh},$out}/share/man/man1/ssh-copy-id.1.gz
```
This could use installManPage.

midchildan replied:
Fixed.

SuperSandro2000 replied:
Okay maybe that wasn't the best suggestion. Your first version was actually better.

SuperSandro2000 commented on pkgs/tools/networking/openssh/copyid.nix:
```diff
@@ -0,0 +1,12 @@
+  };
+} ''
+  install -Dm 755 {${openssh},$out}/bin/ssh-copy-id
+  installManPage ${openssh}/share/man/man1/ssh-copy-id.1.gz
```
```suggestion
  install -Dm 644 {${openssh},$out}/share/man/man1/ssh-copy-id.1.gz
```

SuperSandro2000 committed:
Update pkgs/tools/networking/openssh/copyid.nix

SuperSandro2000 commented on pkgs/tools/networking/openssh/copyid.nix:
```diff
@@ -0,0 +1,12 @@
+{ runCommandNoCC, installShellFiles, openssh }:
+
+runCommandNoCC "ssh-copy-id-${openssh.version}" {
+  nativeBuildInputs = [ installShellFiles ];
+  meta = openssh.meta // {
```
```suggestion
{ runCommandNoCC, openssh }:

runCommandNoCC "ssh-copy-id-${openssh.version}" {
  meta = openssh.meta // {
```

SuperSandro2000 committed:
Update pkgs/tools/networking/openssh/copyid.nix
